### ISSUE TITLE
 fast_xml requires Expat >=2.4.5 (or backported fixes) for secure operation

### ISSUE THREAD

**Issue by hartwork** [processone/fast_xml#46]
Hi fast_xml,

it has come to my attention that the way fast_xml processes element names at… https://github.com/processone/fast_xml/blob/4e8d0ba4ea08d642acf63b25c65f89a6f632ea3b/c_src/fxml_stream.c#L161-L191 …requires Expat >=2.4.5 (with the fix to CVE-2022-25236 from pull request(s) https://github.com/libexpat/libexpat/pull/561 (and https://github.com/libexpat/libexpat/pull/577)) to not be vulnerable to injection of namespace separator characters as demonstrated with vulnerable Expat 2.4.4 by the Dockerfile, below.

If everyone uses fast_xml with Expat >=2.4.5, then fast_xml code does not need any changes. I wanted to be sure you're aware.

For the `Dockerfile`:
```Dockerfile
# Copyright (c) 2022 Sebastian Pipping <sebastian@pipping.org>
# Licensed under the Apache License version 2.0

FROM debian:sid
SHELL ["/bin/bash", "-c"]
RUN apt-get update \
        && \
    apt-get install --no-install-recommends --yes \
            build-essential \
            ca-certificates \
            cmake \
            git \
            make \
        && \
    git clone --depth 1 --branch R_2_4_4 https://github.com/libexpat/libexpat/ 2>/dev/null \
        && \
    cmake -S libexpat/expat \
        && \
    cp libexpat/expat/examples/elements.c{,_BACKUP} \
        && \
    sed "s/XML_ParserCreate(NULL);/XML_ParserCreateNS(NULL, '|');\n  XML_SetReturnNSTriplet(parser, 1);/" -i libexpat/expat/examples/elements.c \
        && \
    make -j$(nproc) \
        && \
    ! diff -u libexpat/expat/examples/elements.c{_BACKUP,} \
        && \
    ./examples/elements <<<'<a><d xmlns="b&#x7C;c" /><d:c xmlns:d="b" /></a>'
```

And for the expected output:

```console
# docker build .
[..]
a
        b|c|d
        b|c|d
[..]
```

Best, Sebastian

**Comment by hartwork** [comment]
@badlop I saw you merged PR #45. Do you have a second? My main point in this^^ ticket was to raise awareness. Should this ticket be closed as "done, everyone is aware" or is there any party that is not yet aware that may need this ticket? Thanks!

**Comment by badlop** [comment]
This library is initially used in ejabberd. When ejabberd is installed from Debian or other operating systems, it will get patched expat libraries... but when installed in other ways, that can't be guaranteed. Also, fast_xml is available as a library and apparently other projects are using it: https://hex.pm/packages/fast_xml and who knows what do those projects use, or any other developers...

It would be great if the vulnerability in expat could be prevented in fast_xml. That way users wouldn't need to update to fixed expat versions. Do you have any clue here, or example in other libraries?

I can't help in fixing the source code... I can only help ensuring this is documented and encouraged when the next version is released.

**Comment by hartwork** [comment]
@badlop thanks for your reply!

> It would be great if the vulnerability in expat could be prevented in fast_xml. That way users wouldn't need to update to fixed expat versions. Do you have any clue here, or example in other libraries?

I'm aware of two ways:
- a) make the build system require libexpat >=2.4.5. Downside would be that people who have the fixes but in the form of backports would then need to bypass or patch the check away, say in fast_xml packaging.
- b) add a small C program to precisely test the needed behavior — backported or not —, and make it part of the build system. More work but more precise.

Both of these would rely on users using the original build system.
I have no strong opinions on any of these, these are just options.

**Comment by hartwork** [comment]
@badlop I was using `'|'` (pipe, hex character reference `&#x7C;`) in my example above but fast_xml is using `'\n'` (line feed, hex character reference `&#xA;`) for a namespace separator. If you replace `&#x7C;` with `&#xA;` in your sample document, you'll get a parse error with Expat >=2.4.5 (or its backported patches) and no error with vulnerable versions.

**Comment by badlop** [comment]
Aha!

Using libexpat1-dev from debian sid (2.4.7-1) this complains:

```
fxml_stream:parse_element(<<"<a><d xmlns='b&#xA;c' /><d:c xmlns:d='b' /></a>">>).
{error,{2,<<"syntax error">>}}
```

But, when using libexpat1-dev from debian stable (2.2.10-2) or from debian testing (2.4.4-1), it gets parsed as:

```
fxml_stream:parse_element(<<"<a><d xmlns='b&#xA;c' /><d:c xmlns:d='b' /></a>">>).
{xmlel,<<"a">>,[],
       [{xmlel,<<"d:c">>,[{<<"xmlns">>,<<"b\nc">>}],[]},
        {xmlel,<<"d:c">>,[{<<"xmlns:d">>,<<"b">>}],[]}]}
```

**Comment by hartwork** [comment]
@badlop yes, but please note that Debian stable — codename `bullseye` — has [Expat `2.2.10-2+deb11u3` with backport fixes](https://tracker.debian.org/news/1310255/accepted-expat-2210-2deb11u3-source-into-stable-security-embargoed-stable-security/) in "channel" `bullseye-security`. So users only have `2.2.10-2` on Debian stable systems that are lacking security updates.

**Comment by hartwork** [comment]
@badlop PS: With regard to testing/bookworm, I believe testing does not have a `-security` channel and is not forwarded to `2.4.7-1` of unstable/sid yet, because of a [migration blocker](https://qa.debian.org/excuses.php?package=expat) that still needs someone to resolve.

**Comment by badlop** [comment]
Ah, right! I forgot to enable bullseye-security, which includes that backported fix, as you explained.

Thanks! I've added a test to the suite, so packagers and admins can verify it easily.

I'll make sure this is mentioned in the changelog when released.

**Comment by hartwork** [comment]
Cool, thank you! :+1:

**Comment by hartwork** [comment]
> I believe testing does not have a `-security` channel and is not forwarded to `2.4.7-1` of unstable/sid yet, because of a [migration blocker](https://qa.debian.org/excuses.php?package=expat) that still needs someone to resolve.

Quick update on that^^: Now resolved, testing/bookworm is at Expat `2.4.7-1` now.
